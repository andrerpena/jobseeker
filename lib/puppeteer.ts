import { ElementHandle, Page } from "puppeteer";
import puppeteer from "puppeteer";

export interface Queryable {
  $x(expression: string): Promise<ElementHandle[]>;
}

export async function getTextFromElement(page: Page, element: ElementHandle) {
  return page.evaluate(element => element.textContent, element);
}

export async function getAttributeFromElement(
  page: Page,
  element: ElementHandle,
  attribute: string
) {
  return page.evaluate(
    (element, attribute) => element.getAttribute(attribute),
    element,
    attribute
  );
}

export async function getInnerHtmlFromElement(
  page: Page,
  element: ElementHandle
) {
  return (await page.evaluate(element => element.innerHTML, element)).trim();
}

export async function getNextElement(page: Page, element: ElementHandle) {
  return page.evaluateHandle(
    element => element.nextElementSibling,
    element
  ) as Promise<ElementHandle>;
}

export async function getElementWithExactText(
  frameBase: Queryable,
  text: string
) {
  const elements = await frameBase.$x(`//*[text()='${text}']`);
  if (elements.length) {
    return elements[0];
  }
  return null;
}

export async function launchPuppeteer(): Promise<puppeteer.Browser> {
  return puppeteer.launch({
    args: ["--no-sandbox", "--disable-setuid-sandbox"]
  });
}
