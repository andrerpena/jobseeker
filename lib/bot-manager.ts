import { BotLogger, Logger } from "./logger";
import colors from "colors";
import puppeteer from "puppeteer";
import { JobInput } from "../graphql-types";
import { addCompany, addJob, getCompany, getJob } from "./graphql-client";
import { getMarkdownFromHtml } from "./markdown";
import { launchPuppeteer } from "./puppeteer";
import { extractLocationTag } from "./location";

const RATE_LIMIT = 30;

export interface LocationDetails {
  raw?: string;
  requiredLocation?: string;
  preferredLocation?: string;
  preferredTimeZone?: number;
  preferredTimeZoneTolerance?: number;
}

export interface SalaryDetails {
  raw?: string;
  exact?: number;
  min?: number;
  max?: number;
  currency?: string;
  equity?: boolean;
}

export interface CompanyDetails {
  displayName: string;
  imageUrl: string;
}

export interface JobDraft {
  link: string;
  draft: any;
}

export interface Bot {
  getName(): string;

  getJobDrafts(
    logger: Logger,
    browser: puppeteer.Browser
  ): Promise<Array<JobDraft | null>>;

  shouldCapture(page: puppeteer.Page): Promise<boolean>;

  getTitle(page: puppeteer.Page, draft: JobDraft): Promise<string>;

  getTags(page: puppeteer.Page, draft: JobDraft): Promise<string[]>;

  getDescriptionHtml(page: puppeteer.Page, draft: JobDraft): Promise<string>;

  getUtcPublishedAt(
    page: puppeteer.Page,
    draft: JobDraft
  ): Promise<Date | null>;

  getLocationDetails(
    page: puppeteer.Page,
    draft: JobDraft
  ): Promise<LocationDetails>;

  getSalaryDetails(
    page: puppeteer.Page,
    draft: JobDraft
  ): Promise<SalaryDetails>;

  getCompany(page: puppeteer.Page, draft: JobDraft): Promise<CompanyDetails>;
}

export class ConsoleBotLogger implements BotLogger {
  private botName: string;

  constructor(botName: string) {
    this.botName = botName;
  }

  private log(color: colors.Color, message: string, data?: any) {
    if (data) {
      console.log(
        color(
          `(Bot: ${this.botName}): ${message} - Data: ${JSON.stringify(
            data,
            null,
            4
          )}`
        )
      );
    } else {
      console.log(color(`(Bot: ${this.botName}) ${message}`));
    }
    return Promise.resolve();
  }

  logError(message: string, data?: any) {
    return this.log(colors.red, message, data);
  }

  logInfo(message: string, data?: any) {
    return this.log(colors.green, message, data);
  }
}

export class BotManager {
  bots: Bot[] = [];
  browserPromise: Promise<puppeteer.Browser>;

  constructor() {
    this.browserPromise = launchPuppeteer();
  }

  register(bot: Bot) {
    this.bots.push(bot);
  }

  async wrapCall<T>(
    call: () => Promise<T>,
    defaultValue: T | null,
    logKey: string,
    url: string,
    logger: BotLogger
  ) {
    try {
      return await call();
    } catch (error) {
      await logger.logError(
        `There was an error processing ${logKey} for URL: ${url}`,
        error
      );
    }
    if (defaultValue) {
      return defaultValue;
    }
    throw new Error(
      `The bot failed to return the ${logKey} value for URL ${url}. Aborting`
    );
  }

  async saveJob(bot: Bot, draft: JobDraft, logger: BotLogger): Promise<any> {
    const browser = await this.browserPromise;
    const page = await browser.newPage();
    logger.logInfo(`Processing: ${draft.link}`);
    try {
      await logger.logInfo("Start processJob");
      await page.goto(draft.link);
      const shouldProceed = await this.wrapCall(
        () => bot.shouldCapture(page),
        true,
        "shouldCapture",
        draft.link,
        logger
      );
      if (!shouldProceed) {
        await logger.logInfo(`Ignoring job: ${draft.link}`);
        return;
      }

      const source = bot.getName();
      const title = await this.wrapCall(
        () => bot.getTitle(page, draft),
        null,
        "getTitle",
        draft.link,
        logger
      );

      const publishedAt = await this.wrapCall(
        () => bot.getUtcPublishedAt(page, draft),
        new Date(),
        "getUtcPublishedAt",
        draft.link,
        logger
      );

      const description = getMarkdownFromHtml(
        await this.wrapCall(
          () => bot.getDescriptionHtml(page, draft),
          "getDescriptionHtml",
          null,
          draft.link,
          logger
        )
      );

      const tags = await this.wrapCall(
        () => bot.getTags(page, draft),
        null,
        "getTags",
        draft.link,
        logger
      );

      if (!tags || tags.length === 0) {
        await logger.logInfo(`Ignoring job because of no tags: ${draft.link}`);
        return;
      }

      const locationDetails = await this.wrapCall<LocationDetails>(
        () => bot.getLocationDetails(page, draft),
        {},
        "getLocationDetails",
        draft.link,
        logger
      );

      const salaryDetails = await this.wrapCall<SalaryDetails>(
        () => bot.getSalaryDetails(page, draft),
        {},
        "getSalaryDetails",
        draft.link,
        logger
      );

      const companyDetails = await this.wrapCall(
        () => bot.getCompany(page, draft),
        null,
        "getCompany",
        draft.link,
        logger
      );

      const getCompanyResult = await getCompany({
        displayName: companyDetails.displayName
      });
      let companyId: string;
      if (getCompanyResult.data.getCompany) {
        companyId = getCompanyResult.data.getCompany.id;
      } else {
        const addCompanyResult = await addCompany({
          input: {
            displayName: companyDetails.displayName,
            imageUrl: companyDetails.imageUrl
          }
        });
        if (!addCompanyResult.data) {
          await logger.logError(
            `could not add company`,
            addCompanyResult.errors
          );
          return;
        }
        companyId = addCompanyResult.data.addCompany.id;
      }

      if (!companyId) {
        throw Error("COMPANY SHOULD NOT BE NULL");
      }

      const locationTag = extractLocationTag(
        locationDetails ? locationDetails.requiredLocation : "",
        title,
        description
      );

      const job: JobInput = {
        title,
        description,
        tags,
        companyId,

        publishedAt: publishedAt
          ? publishedAt.toISOString()
          : new Date().toISOString(),
        url: draft.link,
        locationRequired: locationDetails.requiredLocation,
        locationPreferred: locationDetails.preferredLocation,
        locationPreferredTimezone: locationDetails.preferredTimeZone,
        locationPreferredTimezoneTolerance:
          locationDetails.preferredTimeZoneTolerance,
        locationRaw: locationDetails.raw,
        salaryRaw: salaryDetails.raw,
        salaryExact: salaryDetails.exact,
        salaryMin: salaryDetails.min,
        salaryMax: salaryDetails.max,
        salaryCurrency: salaryDetails.currency,
        salaryEquity: salaryDetails.equity,
        locationTag: locationTag,
        source
      };

      try {
        const result = await addJob({
          job
        });

        if (result.errors) {
          logger.logError(
            `Could not save job link: ${draft.link}. job object: ${job}`,
            result.errors
          );
        } else {
          logger.logInfo(`Successfully saved job`, result.data);
        }
      } catch (error) {
        logger.logError(
          `Could not save job link: ${draft.link}. Error: ${error}`,
          job
        );
      }
    } catch (error) {
      await logger.logError(error, draft);
    } finally {
      await page.close();
      await logger.logInfo("End processJob");
    }
  }

  async run(): Promise<void> {
    const browser = await this.browserPromise;
    for (let bot of this.bots) {
      const logger = new ConsoleBotLogger(bot.getName());
      await logger.logInfo("Start getJobDrafts");
      let drafts = (await bot.getJobDrafts(logger, browser)).filter(
        i => i !== null
      ) as JobDraft[];
      await logger.logInfo("End getJobDrafts");
      await logger.logInfo(`Jobs found: ${drafts.length}`);
      let counter = 0;
      for (let draft of drafts) {
        try {
          if (counter < RATE_LIMIT) {
            logger.logInfo(`counter: ${counter}`);
            const existingJob = await getJob({ jobUrl: draft.link });
            if (existingJob.data.getJob != null) {
              logger.logInfo(
                `Skipping ${draft.link} because it existed already`
              );
              continue;
            }
            const data = await this.saveJob(bot, draft, logger);
            if (data) {
              logger.logInfo(`Processed ${counter} jobs`);
            }
          } else {
            break;
          }
        } finally {
          counter++;
        }
      }
    }
  }
}
