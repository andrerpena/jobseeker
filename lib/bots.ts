import { BotLogger, Logger } from "./logger";
import colors from "colors";
import * as puppeteer from "puppeteer";

export interface Job {
  title: String;
  tags: string[];
  description: string;
  locationDetails?: LocationDetails;
  salaryDetails?: SalaryDetails;
}

export interface LocationDetails {
  raw: string;
  requiredLocation?: string;
  preferredLocation?: string;
  preferredTimeZone?: number;
  preferredTimeZoneTolerance?: number;
}

export interface SalaryDetails {
  raw: string;
  exact?: number;
  min?: number;
  max?: number;
  currency?: string;
  equity?: boolean;
}

export interface JobDraft {
  link: string;
  draft: object;
}

export interface Bot {
  getName(): string;

  getJobDrafts(logger: Logger): Promise<Array<JobDraft | null>>;

  saveJob(job: Job, logger: Logger): Promise<void>;

  shouldCapture(page: puppeteer.Page): Promise<boolean>;

  getTitle(page: puppeteer.Page): Promise<string>;

  getTags(page: puppeteer.Page): Promise<string[]>;

  getDescription(page: puppeteer.Page): Promise<string>;

  getLocationDetails(
    page: puppeteer.Page
  ): Promise<LocationDetails | undefined>;

  getSalaryDetails(page: puppeteer.Page): Promise<SalaryDetails | undefined>;
}

export class ConsoleBotLogger implements BotLogger {
  private botName: string;

  constructor(botName: string) {
    this.botName = botName;
  }

  private log(color: colors.Color, message: string, data?: any) {
    if (data) {
      console.log(
        color(
          `(Bot: ${this.botName}): ${message} - Data: ${JSON.stringify(
            data,
            null,
            4
          )}`
        )
      );
    } else {
      console.log(color(`(Bot: ${this.botName}) ${message}`));
    }
    return Promise.resolve();
  }

  logError(message: string, data?: any) {
    return this.log(colors.red, message, data);
  }

  logInfo(message: string, data?: any) {
    return this.log(colors.green, message, data);
  }
}

export class BotManager {
  bots: Bot[] = [];
  browserPromise: Promise<puppeteer.Browser>;

  constructor() {
    this.browserPromise = puppeteer.launch();
  }

  register(bot: Bot) {
    this.bots.push(bot);
  }

  async processJob(
    bot: Bot,
    draft: JobDraft,
    logger: BotLogger
  ): Promise<void> {
    const browser = await this.browserPromise;
    const page = await browser.newPage();
    try {
      await logger.logInfo("Start processJob");
      await page.goto(draft.link);
      const shouldProceed = await bot.shouldCapture(page);
      if (!shouldProceed) {
        return;
      }

      const title = await bot.getTitle(page);
      const description = await bot.getDescription(page);
      const tags = await bot.getTags(page);
      const locationDetails = await bot.getLocationDetails(page);
      const salaryDetails = await bot.getSalaryDetails(page);

      const job: Job = {
        title,
        description,
        tags,
        locationDetails,
        salaryDetails
      };

      await bot.saveJob(job, logger);
    } catch (error) {
      await logger.logError(error, draft);
    } finally {
      await page.close();
      await logger.logInfo("End processJob");
    }
  }

  async run(): Promise<void> {
    for (let bot of this.bots) {
      const logger = new ConsoleBotLogger(bot.getName());
      await logger.logInfo("Start getJobDrafts");
      let drafts = (await bot.getJobDrafts(logger)).filter(
        i => i !== null
      ) as JobDraft[];
      drafts = drafts.slice(0, 10);
      await logger.logInfo("End getJobDrafts");
      await logger.logInfo(`Jobs found: ${drafts.length}`);
      for (let draft of drafts) {
        await this.processJob(bot, draft, logger);
      }
    }
  }
}
