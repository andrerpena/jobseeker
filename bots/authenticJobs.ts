import {
  Bot,
  CompanyDetails,
  JobDraft,
  LocationDetails,
  SalaryDetails
} from "../lib/bot-manager";
import { BotLogger } from "../lib/logger";
import * as puppeteer from "puppeteer";
import {
  getAttributeFromElement,
  getInnerHtmlFromElement,
  getTextFromElement
} from "../lib/puppeteer";
import { getMarkdownFromHtml, removeMarkdown } from "../lib/markdown";
import { extractTags } from "../lib/tag-extractor";

export class AuthenticJobs implements Bot {
  buildAbsoluteUrl(relativeUrl: string) {
    return `https://authenticjobs.com${relativeUrl}`;
  }

  extractLocationDetails(remoteDetails: string): LocationDetails {
    throw new Error("Not implemented");
  }

  extractSalaryDetails(salary: string): SalaryDetails {
    throw new Error("Not implemented");
  }

  async shouldCapture(page: puppeteer.Page): Promise<boolean> {
    return true;
  }

  async getTitle(page: puppeteer.Page): Promise<string> {
    const jobDetailsHeader = await page.$(
      "div#content .row.clear-after:first-child"
    );
    if (!jobDetailsHeader) {
      throw new Error("jobDetailsHeader was not supposed to be null");
    }
    const titleElement = await jobDetailsHeader.$(".column h1");
    if (!titleElement) {
      throw new Error("titleElement was not supposed to be null");
    }
    return (await getTextFromElement(page, titleElement)).trim();
  }

  async getTags(page: puppeteer.Page): Promise<string[]> {
    const descriptionHtml = await this.getDescriptionHtml(page);
    const description = removeMarkdown(getMarkdownFromHtml(descriptionHtml));
    return extractTags(description).map(t => t.name);
  }

  async getDescriptionHtml(page: puppeteer.Page): Promise<string> {
    const description = await page.$("#the_listing .description");
    const html = await getInnerHtmlFromElement(page, description);
    if (!html) {
      throw new Error("html was not supposed to be null");
    }
    return html;
  }

  async getLocationDetails(page: puppeteer.Page): Promise<LocationDetails> {
    throw new Error("Not implemented");
  }

  async getSalaryDetails(page: puppeteer.Page): Promise<SalaryDetails> {
    throw new Error("Not implemented");
  }

  async getJobDrafts(
    logger: BotLogger,
    browser: puppeteer.Browser
  ): Promise<Array<JobDraft | null>> {
    const page = await browser.newPage();
    await page.goto("https://authenticjobs.com/#remote=true");
    const jobs = await page.$$(".listedJob a.listing");
    return Promise.all(
      jobs.map(async jobElement => {
        const href = await getAttributeFromElement(page, jobElement, "href");
        return {
          link: this.buildAbsoluteUrl(href),
          draft: null
        };
      })
    );
  }

  getName(): string {
    return "authentic-jobs";
  }

  async getCompany(page: puppeteer.Page): Promise<CompanyDetails> {
    throw new Error("Not implemented");
  }

  async getUtcPublishedAt(
    page: puppeteer.Page,
    draft: JobDraft
  ): Promise<Date | null> {
    const header = await page.$("#the_listing time");
    const dateString = await getAttributeFromElement(page, header, "datetime");
    const d = new Date(0); // The 0 there is the key, which sets the date to the epoch
    d.setUTCSeconds(dateString);
    return d;
  }
}
