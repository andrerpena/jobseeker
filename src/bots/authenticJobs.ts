import { Bot, CompanyDetails, JobDraft, SalaryDetails } from "../bot-manager";
import { BotLogger } from "../logger";
import * as puppeteer from "puppeteer";
import {
  getAttributeFromElement,
  getInnerHtmlFromElement,
  getTextFromElement
} from "../puppeteer";
import { getMarkdownFromHtml, removeMarkdown } from "../markdown";
import { extractTags } from "../tag-extractor";
import { LocationDetailsInput } from "../../graphql-types";
import { extractLocation } from "../location";

export class AuthenticJobs implements Bot {
  buildAbsoluteUrl(relativeUrl: string) {
    return `https://authenticjobs.com${relativeUrl}`;
  }

  getLocationFromText(text: string) {
    if (text) {
      const processedText = text;
      if (
        processedText === "Anywhere in the world" ||
        processedText === "Anywhere"
      ) {
        return null;
      }
      const match = processedText.match(/Anywhere in (.*)/);
      if (match) {
        return match[1].trim();
      }
      return processedText;
    }
    return null;
  }

  async shouldCapture(page: puppeteer.Page): Promise<boolean> {
    const remoteIcon = await page.$(".ss-wifi");
    return !!remoteIcon;
  }

  async getTitle(page: puppeteer.Page): Promise<string> {
    const jobDetailsHeader = await page.$(
      "div#content .row.clear-after:first-child"
    );
    if (!jobDetailsHeader) {
      throw new Error("jobDetailsHeader was not supposed to be null");
    }
    const titleElement = await jobDetailsHeader.$(".column h1");
    if (!titleElement) {
      throw new Error("titleElement was not supposed to be null");
    }
    return (await getTextFromElement(page, titleElement)).trim();
  }

  async getTags(page: puppeteer.Page): Promise<string[]> {
    const descriptionHtml = await this.getDescriptionHtml(page);
    const title = await this.getTitle(page);
    const description = removeMarkdown(getMarkdownFromHtml(descriptionHtml));
    return extractTags(title + " " + description).map(t => t.name);
  }

  async getDescriptionHtml(page: puppeteer.Page): Promise<string> {
    const description = await page.$("#the_listing .description");
    const html = await getInnerHtmlFromElement(page, description);
    if (!html) {
      throw new Error("html was not supposed to be null");
    }
    return html;
  }

  async getLocationDetails(
    page: puppeteer.Page
  ): Promise<LocationDetailsInput> {
    const description = getMarkdownFromHtml(
      await this.getDescriptionHtml(page)
    );
    const result: LocationDetailsInput =
      extractLocation(description, true) || {};

    let regionElement = await page.$("#location a");
    if (!regionElement) {
      regionElement = await page.$(".ss-location");
    }
    if (regionElement) {
      let locationRaw = await getTextFromElement(page, regionElement);
      if (locationRaw) {
        locationRaw = locationRaw.trim().replace(/(\r\n|\n|\r)/gm, "");
      }
      const location = this.getLocationFromText(locationRaw);
      const extractedLocation = location
        ? extractLocation(location, false)
        : {};

      return {
        ...result,
        description: locationRaw,
        ...(extractedLocation || {})
      };
    }
    return result;
  }

  async getSalaryDetails(page: puppeteer.Page): Promise<SalaryDetails> {
    return {};
  }

  async getJobDrafts(
    logger: BotLogger,
    browser: puppeteer.Browser
  ): Promise<Array<JobDraft | null>> {
    const page = await browser.newPage();
    await page.goto("https://authenticjobs.com/#remote=true");
    const jobs = await page.$$(".listedJob a.listing");
    return Promise.all(
      jobs.map(async jobElement => {
        const href = await getAttributeFromElement(page, jobElement, "href");
        return {
          link: this.buildAbsoluteUrl(href),
          draft: null
        };
      })
    );
  }

  getName(): string {
    return "authentic-jobs";
  }

  async getCompany(page: puppeteer.Page): Promise<CompanyDetails> {
    const overview = await page.$("#the_company header");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const companyNameElement = await overview.$(".headings h1");
    if (!companyNameElement) {
      throw new Error("companyNameElement was not supposed to be null");
    }
    const companyName = await getTextFromElement(page, companyNameElement);

    if (!companyName) {
      throw new Error("companyName was not supposed to be null");
    }

    const companyImageLink = await overview.$("a img");

    const companyImageUrl = companyImageLink
      ? await getAttributeFromElement(page, companyImageLink, "src")
      : "";

    return {
      displayName: companyName,
      imageUrl: companyImageUrl
    };
  }

  async getUtcPublishedAt(
    page: puppeteer.Page,
    draft: JobDraft | null
  ): Promise<Date | null> {
    const header = await page.$("#the_listing time");
    const dateString = await getAttributeFromElement(page, header, "datetime");
    const d = new Date(0); // The 0 there is the key, which sets the date to the epoch
    d.setUTCSeconds(dateString);
    return d;
  }
}
