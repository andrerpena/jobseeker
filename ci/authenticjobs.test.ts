import puppeteer from "puppeteer";
import { ConsoleBotLogger } from "../lib/bot-manager";
import { launchPuppeteer } from "../lib/puppeteer";
import { AuthenticJobs } from "../bots/authenticJobs";

const authenticJobs = new AuthenticJobs();

const JOB_REMOTE_URL =
  "https://authenticjobs.com/jobs/31311/senior-front-end-developer";

const JOB_REMOTE_URL_US =
  "https://authenticjobs.com/jobs/31264/senior-front-end-engineer-ember-js";

const JOB_URL_ANYWHERE =
  "https://authenticjobs.com/jobs/31308/copywriter-with-an-seo-emphasis";

const JOB_URL_ANYWHERE_IN_THE_WORLD =
  "https://authenticjobs.com/jobs/31269/web-developer";

describe("AuthenticJobs", () => {
  jest.setTimeout(30000);

  let browser: puppeteer.Browser;
  beforeAll(async () => {
    browser = await launchPuppeteer();
  });
  afterAll(async () => {
    return browser.close();
  });
  describe("drafts", () => {
    it("should work", async () => {
      const jobs = await authenticJobs.getJobDrafts(
        new ConsoleBotLogger("authentic-jobs"),
        browser
      );
      expect(jobs.length > 0);
      expect(jobs).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            link: expect.any(String),
            draft: null
          })
        ])
      );
    });
  });
  describe("getTitle", () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL);
      const title = await authenticJobs.getTitle(page);
      expect(title).toBe("Senior Front EndÂ Developer");
    });
  });
  describe("getDescriptionHtml", async () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL);
      const description = await authenticJobs.getDescriptionHtml(page);
      expect(description).toMatchSnapshot();
    });
  });
  describe("getTags", async () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL);
      const tags = await authenticJobs.getTags(page);
      expect(tags).toEqual([
        "reactjs",
        "node.js",
        "javascript",
        "html",
        "css",
        "google-maps",
        "sql",
        "elasticsearch",
        "amazon-web-services",
        "jenkins",
        "ios",
        "react-native"
      ]);
    });
  });
  describe("getUtcPublishedAt", async () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL);
      const date = await authenticJobs.getUtcPublishedAt(page, null);
      expect(date).toEqual(new Date("2019-04-11T19:46:03.000Z"));
    });
  });
  describe("getLocationDetails", () => {
    it("should work for USA", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_US);
      const companyDetails = await authenticJobs.getLocationDetails(page);
      expect(companyDetails).toEqual({
        raw: "Anywhere in United States",
        requiredLocation: "United States"
      });
    });
    it("should work for Anywhere", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_URL_ANYWHERE);
      const companyDetails = await authenticJobs.getLocationDetails(page);
      expect(companyDetails).toEqual({
        raw: "Anywhere",
        requiredLocation: null
      });
    });
    it("should work for Anywhere in the World", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_URL_ANYWHERE_IN_THE_WORLD);
      const companyDetails = await authenticJobs.getLocationDetails(page);
      expect(companyDetails).toEqual({
        raw: "Anywhere in the world",
        requiredLocation: null
      });
    });
  });
});
