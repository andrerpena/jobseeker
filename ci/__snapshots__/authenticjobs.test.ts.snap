// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticJobs getDescriptionHtml should work 1`] = `
"<div id=\\"posted\\" class=\\"ss-gizmo ss-stopwatch\\"><time datetime=\\"1555011963\\">Posted Yesterday</time> – Accepting applications</div>
    
    <p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Who We Are</strong></p>

<p></p><p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">At Outside Analytics we are passionate about the outdoors and building software that increases outdoor accessibility, usage, and understanding. We are building a data analytics and geospatial processing platform with an innovative web and mobile interface to empower people’s outdoor experiences. We believe that the right technology can create stronger connections between people and nature, improve health and wellness, increase environmental empathy, and drive support for public land conservation.</p>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Position Responsibilities</strong></p>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">Development, test and deployment of complex and highly performant web applications. Assist in the development of strategic and technology roadmaps, and support the development of mobile applications.</p>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Required Qualifications and Skills</strong></p>
<ul style=\\"box-sizing: border-box; margin: 0px 0px 10px 45px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Passionate about the outdoors and outdoor recreation</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Expertise in React, Node, JavaScript, HTML, CSS</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Proficiency working with mapping libraries and APIs such as Mapbox or Google Maps</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">API development and data driven RESTful web services</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Performant querying of backend data stores such as SQL databases and Elasticsearch</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Experience developing and deploying into Amazon Web Services (AWS)</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">DevOps practices and development tool chains (Github, Jenkins, Chef, etc)</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Interpersonal skills, taking accountability for objectives, working with minimal oversight</li>
</ul>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Additional Qualifications and Skills</strong></p>
<ul style=\\"box-sizing: border-box; margin: 0px 0px 10px 45px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Mobile development (iOS, Android, React Native)</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Experience with 3D mapping technology (Cesium, Google Earth, etc.)</li>
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Familiarity with data science</li>
</ul>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Education and Training Required</strong></p>
<ul style=\\"box-sizing: border-box; margin: 0px 0px 10px 45px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">4 year college (Bachelor’s Degree) or higher. Computer Science degree preferred.</li>
</ul>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Minimum Years of Experience</strong></p>
<ul style=\\"box-sizing: border-box; margin: 0px 0px 10px 45px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">4+ years of real world experience developing proven applications and systems.</li>
</ul>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Location</strong></p>
<ul style=\\"box-sizing: border-box; margin: 0px 0px 10px 45px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Boulder, Colorado</li>
</ul>
<p style=\\"box-sizing: border-box; margin: 0px 0px 10px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\"><strong style=\\"box-sizing: border-box;\\">Other</strong></p>
<ul style=\\"box-sizing: border-box; margin: 0px 0px 10px 45px; padding: 0px; color: #2d2d2d; font-family: HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 14px; background-color: #ffffff;\\">
<li style=\\"box-sizing: border-box; margin: 0px; padding: 0px 10px 0px 0px; list-style: disc; line-height: 21px; overflow-wrap: break-word;\\">Must be US Citizen</li>
</ul><p></p>"
`;
